Add DNS-schema Specification

johnbelamaric opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://github.com/kubernetes/kubernetes/blob/master/CONTRIBUTING.md and developer guide https://github.com/kubernetes/kubernetes/blob/master/docs/devel/development.md
2. If you want *faster* PR reviews, read how: https://github.com/kubernetes/kubernetes/blob/master/docs/devel/faster_reviews.md
3. Follow the instructions for writing a release note: https://github.com/kubernetes/kubernetes/blob/master/docs/devel/pull-requests.md#release-notes
-->

**What this PR does / why we need it**:
This PR establishes an explicit baseline DNS schema for Kubernetes. Currently, this information is contained in a few places in the documentation. This PR gathers this into one place and adds additional, explicit detail to what the supported queries and expected responses are. The intention is to provide a common schema that different DNS service discovery implementations can implement.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: None

**Special notes for your reviewer**:
This started life here https://docs.google.com/document/d/1B4ZA1HwprWMzyT7nUZLAZ93Fjwcdt9qs0aXfifKxpTs/edit# but this PR is the result. Currently kube-dns does not support this schema 100%. There are a few items in here that are not yet in kube-dns. I will add comments to them on the PR. This makes the spec more aspirational than current. We may want to note that directly in the spec until kube-dns fully meets it (or change the spec, of course).

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access) 
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`. 
-->
```NONE
```

johnbelamaric committed:
Add DNS-schema Specification

Add a DNS-schema spec to create a baseline for any DNS-based
service discovery component to follow.

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/38408)
<!-- Reviewable:end -->

johnbelamaric commented on cmd/kube-dns/SPEC.md:
```diff
@@ -0,0 +1,99 @@
+
+  - If a Pod exists with IP address `<a>.<b>.<c>.<d>`, the result will be an A record with the specified name and the IP address `<a>.<b>.<c>.<d>`.
+
+  - If no such Pod exists, the result will be `NXDOMAIN`.
```
Current kube-dns behavior is to ignore the state of the cluster and simply reply to every query of this form with <a>.<b>.<c>.<d>, whether anything is at that IP or not.

johnbelamaric replied:
@thockin I believe that it is important to respond based on the state of the cluster. My understanding of the use case for the existing `<ip>.pod.cluster.local` responses is to enable a wildcard certificate like *.ns.pod.cluster.local. However, if the DNS always returns the IP regardless of the state of the cluster, then you are weakening the guarantee of identify provided by certificate.

In cases where a cert is being used like this, there must be an internal application-level discovery mechanism being used to tell clients about the various pods that are running as servers. This is the only way the clients would know about the DNS name like `a-b-c-d.default.pod.cluster.local`. So, consider the situation in which someone is able to subvert that discovery mechanism. Suppose I have a namespace, `goodguy` with a set of pods and I get a signed wildcard certificate `*.goodguy.pod.cluster.local`. Now, there is someone nefarious on my cluster who launches a malicious pod with IP w.x.y.z in a different namespace (say `badguy`). If the DNS reflects the real state of the cluster, then the bad guy would have to push my IP into the discovery mechanism as `w-x-y-z.badguy.pod.cluster.local`, and so the certificate validation would fail, and our clients would ignore that pod. However, if the DNS always echoes back the IP without validating the cluster state, then the bad guy can push `w-x-y-z.goodguy.pod.cluster.local` into the discovery mechanism, and still have the clients reach the malicious pod. The certificate will be seen as valid and matching the server host name.

johnbelamaric replied:
@bowei any thoughts on this?

thockin commented on cmd/kube-dns/SPEC.md:
```diff
@@ -0,0 +1,99 @@
+
+  - For endpoints, in order of precedence:
+    - The value of the endpoint’s `hostname` field.
+    - The value of the `HostName` field stored in the map indexed by the endpoint’s IP address in the `endpoints.beta.kubernetes.io/hostnames-map` annotation.
```
I don't think it's really worth documenting this.  The beta needs to go away O(soon)

johnbelamaric replied:
Ok, simpler is better.

thockin commented on cmd/kube-dns/SPEC.md:
```diff
@@ -0,0 +1,99 @@
+
+  - If no Service named `<service>` exists in `<ns>`, the result will be `NXDOMAIN`.
+
+  - If a Service named `<service>` exists in `<ns>` with a ClusterIP, and the service has a port named `<port>` using protocol `<proto>`, the result in the Answer section will be a single SRV record with the port number of the selected port and the name of the A record for the service (`<service>.<ns>.svc.<zone>`). If the service has exactly one port and it is unnamed, then `<port>` may be empty. ...
```
This block starts with "If a Service named `<service>` ... has a port named `<port>`" and then later says "If the service has exactly one port and it is unnamed".

I don't think the same block should handle named and unnamed.  And anyway, we can't really do unnamed SRV, can we?  I thought _port is required.  Is `_tcp.svc.ns.zone` supposed to work?

johnbelamaric replied:
No, I don't think that is supposed to work. I can remove SRV for unnamed ports.

bowei replied:
I'm pretty sure if port is not specified, the entry is not generated in the current implementation

thockin commented on cmd/kube-dns/SPEC.md:
```diff
@@ -0,0 +1,99 @@
+
+  - If a Service named `<service>` exists in `<ns>` with a ClusterIP, and the service has a port named `<port>` using protocol `<proto>`, the result in the Answer section will be a single SRV record with the port number of the selected port and the name of the A record for the service (`<service>.<ns>.svc.<zone>`). If the service has exactly one port and it is unnamed, then `<port>` may be empty. ...
+
+  - If a Service named `<service>` exists in `<ns>` without a ClusterIP, and the Service has no ready endpoints, the result will an empty Answer section.
```
missing a "be" in "will be"

thockin commented on cmd/kube-dns/SPEC.md:
```diff
@@ -0,0 +1,99 @@
+
+- **Type:** TXT, **Name:** `dns-version.<zone>`
+
+  - Returns the semantic versioning of the DNS-schema in use in this cluster. For example, “1.0”.
```
nit: 1.0 is not semver - semver requires 3 fields.

thockin commented:
I don't think this is the right home for this spec-doc.  Maybe @devin-donnelly has a better idea for where to store these sorts of docs?

bowei commented:
Suggestion -- I would render the resulting DNS entries in the canonical format given in the RFCs and by `dig` and other tools (e.g. `_http._tcp.ns.svc.c.local. 30  IN   SRV  10 100 9876 server1.foo.ns.svc.c.local`). This will result in no ambiguity regarding the required result. Right now the description is in prose...

miekg commented:
As commented on the google doc... does it make sense to also give the the path and contents of what is written in etcd? (i.e. SkyDNS' `msg.Service`)

johnbelamaric commented:
@miekg I would rather keep it agnostic of the implementation.

johnbelamaric commented:
Sorry for all the pushes. My Markdown preview editor is NOT accurate...

johnbelamaric commented on cmd/kube-dns/SPEC.md:
```diff
@@ -0,0 +1,140 @@
+    - Answer Example:
+      - `my-pet.my-service.default.svc.cluster.local. 4 IN A 10.3.0.100`
+- **Question:** `<a>-<b>-<c>-<d>.<ns>.pod.<zone>. IN A`
+  - If a Pod exists with IP address `<a>.<b>.<c>.<d>`, the result will be an A record with the specified name and the IP address `<a>.<b>.<c>.<d>`.
```
As stated on an earlier commit, this is not how kube-dns does it. But I don't think how it is done today is the most secure; see the comment on the first push of this file.

bowei replied:
Agreed it is not secure, my reading of it was to avoid the hassle of issuing lots of certificates. We can try deprecating it as the use case is somewhat strange.

johnbelamaric commented on cmd/kube-dns/SPEC.md:
```diff
@@ -0,0 +1,140 @@
+    - Answer Format:
+       - `_<port>._<proto>.<service>.<ns>.svc.<zone>. <ttl> IN SRV <weight> <priority> <port-number> <service>.<ns>.svc.<zone>.`
+    - Answer Example:
+       - `_https._tcp.kubernetes.default.svc.cluster.local. 30 IN SRV 10 100 443 kubernetes.default.svc.cluster.local.`
```
Currently kube-dns sticks an arbitrary identifier in front of the `<service>.<ns>...` name for the A record.

bowei replied:
Is the identifier not the hash you have in the hostname definition above?

johnbelamaric commented on cmd/kube-dns/SPEC.md:
````diff
@@ -0,0 +1,140 @@
+    - Answer Example:
+```
+      _https._tcp.headless.default.svc.cluster.local. 4 IN SRV 10 100 443 my-pet-1.headless.default.svc.cluster.local.
+      _https._tcp.headless.default.svc.cluster.local. 4 IN SRV 10 100 443 my-pet-2.headless.default.svc.cluster.local.
````
I don't **think** kube-dns will use the _hostname_ today - instead it is always an arbitrary identifier.

bowei replied:
See above.

johnbelamaric commented on cmd/kube-dns/SPEC.md:
````diff
@@ -0,0 +1,140 @@
+      _https._tcp.my-service.default.svc.cluster.local.  4 IN SRV 10 100 443 my-service.default.svc.cluster.local.
+      _http._tcp.my-service.default.svc.cluster.local.   4 IN SRV 10 100 80  my-service.default.svc.cluster.local.
+      _syslog._udp.my-service.default.svc.cluster.local. 4 IN SRV 10 100 514 my-service.default.svc.cluster.local.
+```
````
Ditto - I think kube-dns puts an arbitrary identifier before the service name.

johnbelamaric commented on cmd/kube-dns/SPEC.md:
````diff
@@ -0,0 +1,140 @@
+      _https._tcp.headless.default.svc.cluster.local.  4 IN SRV 10 100 443 438934893.headless.default.svc.cluster.local.
+      _http._tcp.headless.default.svc.cluster.local.   4 IN SRV 10 100 80  438934893.headless.default.svc.cluster.local.
+      _syslog._udp.headless.default.svc.cluster.local. 4 IN SRV 10 100 514 438934893.headless.default.svc.cluster.local.
+```
````
kube-dns doesn't do this. The LHS here instead would all just be like the question `headless.default.svc.cluster.local.` which means you don't know which port/IP combo goes with each service and protocol. It also puts on the arbitrary identifier.

```
/ # dig headless.default.svc.cluster.local. srv

; <<>> DiG 9.10.4-P3 <<>> headless.default.svc.cluster.local. srv
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 27032
;; flags: qr aa rd ra; QUERY: 1, ANSWER: 5, AUTHORITY: 0, ADDITIONAL: 5

;; QUESTION SECTION:
;headless.default.svc.cluster.local. IN	SRV

;; ANSWER SECTION:
headless.default.svc.cluster.local. 30 IN SRV	10 20 0 3862303234343538.headless.default.svc.cluster.local.
headless.default.svc.cluster.local. 30 IN SRV	10 20 0 6533653966373130.headless.default.svc.cluster.local.
headless.default.svc.cluster.local. 30 IN SRV	10 20 0 6463313736393962.headless.default.svc.cluster.local.
headless.default.svc.cluster.local. 30 IN SRV	10 20 0 6638383363643661.headless.default.svc.cluster.local.
headless.default.svc.cluster.local. 30 IN SRV	10 20 0 6337356433333633.headless.default.svc.cluster.local.

;; ADDITIONAL SECTION:
3862303234343538.headless.default.svc.cluster.local. 30	IN A 10.2.10.72
6533653966373130.headless.default.svc.cluster.local. 30	IN A 10.2.6.51
6463313736393962.headless.default.svc.cluster.local. 30	IN A 10.2.6.52
6638383363643661.headless.default.svc.cluster.local. 30	IN A 10.2.10.70
6337356433333633.headless.default.svc.cluster.local. 30	IN A 10.2.10.71

;; Query time: 2 msec
;; SERVER: 10.3.0.10#53(10.3.0.10)
;; WHEN: Wed Dec 14 20:48:41 UTC 2016
;; MSG SIZE  rcvd: 317
```

johnbelamaric reviewed (commented):
I believe this addresses all previous comments.

bowei commented on cmd/kube-dns/SPEC.md:
```diff
@@ -0,0 +1,140 @@
+       - `_<port>._<proto>.<service>.<ns>.svc.<zone>. <ttl> IN SRV <weight> <priority> <port-number> <service>.<ns>.svc.<zone>.`
+    - Answer Example:
+       - `_https._tcp.kubernetes.default.svc.cluster.local. 30 IN SRV 10 100 443 kubernetes.default.svc.cluster.local.`
+  - If a Service named `<service>` exists in `<ns>` without a ClusterIP, and the Service has no ready endpoints, the result will be an empty Answer section.
```
rcode = 0

bowei commented on cmd/kube-dns/SPEC.md:
````diff
@@ -0,0 +1,140 @@
+      _https._tcp.headless.default.svc.cluster.local. 4 IN SRV 10 100 443 my-pet-2.headless.default.svc.cluster.local.
+      _https._tcp.headless.default.svc.cluster.local. 4 IN SRV 10 100 443 438934893.headless.default.svc.cluster.local.
+```
+- **Question:** `<service>.<ns>.svc.<zone>. IN SRV`
````
I'm not sure about this set of records, as it doesn't follow the SRV RFC name format of _service._protocol. It might be a mistake. I would prefer to not have this if it is not used. I am going to ask the community to see if people are using it.

bowei commented on cmd/kube-dns/SPEC.md:
```diff
@@ -0,0 +1,140 @@
+    - Answer Example:
+      - `dns-version.cluster.local. 28800 IN TXT "1.0.0"`
+
+## Schema Extensions
```
We should probably go into more detail about extensions cannot overlap their domain names with the names given above? Otherwise, you will have incompatible extensions.
Some restrictions off the top of my head:
- For a given name, cannot have additional records be present for types specified by this doc (e.g. you cannot stick more A records where A records already exist).
- For a given class of names (e.g. <service>.<ns>.svc.<cluster>), extensions cannot have records that intersect with that set.

bowei commented on cmd/kube-dns/SPEC.md:
```diff
@@ -0,0 +1,140 @@
+    - Answer Format:
+      - `<hostname>.<service>.<ns>.svc.<zone>. <ttl> IN A <endpoint-ip>`
+    - Answer Example:
+      - `my-pet.my-service.default.svc.cluster.local. 4 IN A 10.3.0.100`
```
Given the hash hostname in the definition above, you should give a hash hostname example as well, as it will match the A record that SRV records `target` field refers to below without special casing.

bowei commented on cmd/kube-dns/SPEC.md:
```diff
@@ -0,0 +1,140 @@
+`<ns>` = a Namespace
+
+All comparisons between query names and data in Kubernetes is **case-insensitive**.
+
```
separate section for Definitions?

bowei reviewed (changes requested):
Sorry for the late reply.

Great doc

I think we might want to organize by records instead of by query/response. See my comment inline below.

johnbelamaric commented on cmd/kube-dns/SPEC.md:
```diff
@@ -0,0 +1,140 @@
+
+This document describes version 1.0.0 of the schema.
+
+## Query Responses
```
I can do this - it's almost ready. It does change the spec a bit. For example, querying for SRV without _port._proto will no longer return anything (at least, not according to the spec). Assuming we are OK with that sort of change, then I think this makes sense.

johnbelamaric committed:
Rewrite describing records that must exist

Re-wrote the specification to indicate the records that must exist.
This change the behavior of a few queries that were acting something
like wildcards. For example, querying for an SRV record with the
service name and no _port._proto prefix is no longer required to
return any records.

Other minor updates to link to external references and correct typos.

johnbelamaric committed:
Fix formatting of headless SRV answer example.

johnbelamaric commented:
Ok, I  re-wrote the specification to indicate the records that must exist. This changes the behavior of a few queries that were acting something like wildcards. For example, querying for an SRV record with the service name and no _port._proto prefix is no longer required to return any records.

I also added a deprecated records section describing the current `pod` query behavior.

johnbelamaric committed:
Add reference to rcode

thockin commented on cmd/kube-dns/SPEC.md:
```diff
@@ -0,0 +1,184 @@
+- Record Format:
+  - `<a>-<b>-<c>-<d>.<ns>.pod.<zone>. <ttl> IN A <a>.<b>.<c>.<d>`
+
+This behavior is deprecated and not required to satisfy this specification.
```
We can't be quite so cavalier about deprecation.  Someone out there is using this.

johnbelamaric replied:
Ok. How do you suggest we proceed? Announce as deprecated in 1.0.0 and remove in 2.0.0?

smarterclayton replied:
We should announce an intent to deprecate it in favor of named headless services and take feedback.  All use cases I'm aware of are headless like things or software that requires reverse lookup of a record to match the local IP, which named headless should solve.

thockin commented:
@k8s-bot ok to test

thockin commented:
I'd love to get this merged ASAP. 

@devin-donnelly we need guidance on where you want "spec-like" docs to go.

* will be maintained as the source of truth for the exact interface of a critical surface of the system
* not aimed at end-users as much as implementors.

thockin commented:
This is interesting.  The deprecation policy only addresses parts of the
core system and "main" API.  While kube-dns, per se, is NOT part of the
core, this spec more or less *is*.  While we can version the spec, users
don't get to choose which DNS spec they use, so the spec version is sort of
irrelevant.

1) We could couple it to kube API versions and say that this is part of the
v1 kube API, and follow the same deprecation timeline.

2) We could ignore the kube API but follow the 1 year deprecation warning.

3) We could add a way to run both, and let end users choose which DNS spec
they want.

(3) is the most work by far, and probably most correct, but I think this is
not worth the energy.

I'd say we follow (1) unless we are willing to invest the energy to hunt
for anyone who uses this (such as the original poster) and make wide and
broad announcements, in which case (2) might be OK.


On Tue, Dec 27, 2016 at 5:28 PM, John Belamaric <notifications@github.com>
wrote:

> *@johnbelamaric* commented on this pull request.
> ------------------------------
>
> In cmd/kube-dns/SPEC.md
> <https://github.com/kubernetes/kubernetes/pull/38408>:
>
> > +Given a _ready_ endpoint with _hostname_ of `<hostname>` and IP address
> +`<a>.<b>.<c>.<d>`, a `PTR` record of the following form must exist.
> +- Record Format:
> +  - `<d>.<c>.<b>.<a>.in-addr.arpa. <ttl> IN PTR <hostname>.<service>.<ns>.svc.<zone>.`
> +- Question Example:
> +  - `100.0.3.10.in-addr.arpa. IN PTR`
> +- Answer Example:
> +  - `100.0.3.10.in-addr.arpa. 14 IN PTR my-pet.headless.default.svc.cluster.local.`
> +
> +### Deprecated Records
> +Kube-DNS versions prior to implementation of this specification also replied
> +with an `A` record of the form below for any values of `<a>`, `<b>`, `<c>`, and `<d>` between 0 and 254:
> +- Record Format:
> +  - `<a>-<b>-<c>-<d>.<ns>.pod.<zone>. <ttl> IN A <a>.<b>.<c>.<d>`
> +
> +This behavior is deprecated and not required to satisfy this specification.
>
> Ok. How do you suggest we proceed? Announce as deprecated in 1.0.0 and
> remove in 2.0.0?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/38408>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVIK7HYvdl58u60KzaJcwQrUvhQxXks5rMbtBgaJpZM4LIPC2>
> .
>

smarterclayton commented on cmd/kube-dns/SPEC.md:
```diff
@@ -0,0 +1,184 @@
+
+## Schema Extensions
+Specific implementations may choose to extend this schema, but the RRs in this
+document must be a subset of the RRs produced by the implementation.
```
Support for federation documented here or elsewhere?

johnbelamaric replied:
Next version. One step at a time.

smarterclayton commented:
General note - anyone signing for `*.pod.*` should be warned not to do it.  We probably need a security recommendations document somewhere for what to sign.

bowei commented:
one slight tweak -- it would be useful to number the items so it is easy for people to identify which section of the spec is being discussed/validated/etc

johnbelamaric committed:
Add numbering, cleanup non-ASCII characters, address deprecation warning

kargakis commented:
Fixes https://github.com/kubernetes/kubernetes/issues/32381

bowei commented:
@thockin -- did @devin-donnelly get back to you on the spec location? It would be great to get this first draft committed in a repo. IMHO probably makes sense to put it in kubernetes/dns/doc/spec, unless we put all specification "stuff" together, rather than organize by subject area.

thockin commented:
I was thinking github.kubernetes.io repo in docs/specs, but I guess before
we do that we want some critical mass of specs?  If we want to get this
merged, dns/docs/spec.md or just dns/SPEC.md is ok for now.

On Mon, Jan 9, 2017 at 9:50 PM, Bowei Du <notifications@github.com> wrote:

> @thockin <https://github.com/thockin> -- did @devin-donnelly
> <https://github.com/devin-donnelly> get back to you on the spec location?
> It would be great to get this first draft committed in a repo. IMHO
> probably makes sense to put it in kubernetes/dns/doc/spec, unless we put
> all specification "stuff" together, rather than organize by subject area.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/38408#issuecomment-271492566>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVF8XdM2AemESryqLnnoqqiSmwDkPks5rQxwpgaJpZM4LIPC2>
> .
>

bowei commented:
hmm -- I don't see a specs folder in the website. I will commit this to the kubernetes/dns repo for the time being.

thockin commented:
That was a proposal for Devin to kill or not.

On Mon, Jan 9, 2017 at 10:22 PM, Bowei Du <notifications@github.com> wrote:

> hmm -- I don't see a specs folder in the website. I will commit this to
> the kubernetes/dns repo for the time being.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/38408#issuecomment-271496362>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVERSvsyeKExJf5V5J-GZOn2N8Mhiks5rQyOsgaJpZM4LIPC2>
> .
>

bowei commented:
@thockin @johnbelamaric -- I opened a new PR against dns, please review https://github.com/kubernetes/dns/pull/24

bowei commented:
merged in dns!

smarterclayton commented:
Thanks for sticking with this! :)

On Jan 10, 2017, at 1:13 PM, Bowei Du <notifications@github.com> wrote:

merged in dns!

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/38408#issuecomment-271652906>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/ABG_p-l_7JQeEOzQUDANmSCuLoB1kqhnks5rQ8pggaJpZM4LIPC2>
.

bowei commented:
Thanks John for the work -- feel free to close this PR.

johnbelamaric commented:
Closing as this has merged in the new dns repository.
